### PR TITLE
Container resource creation using ARM templates

### PR DESCRIPTION
**Pull request by ds-ms**

### CODE REVIEW

**Comment by SumiranAgg** [comment]
I would suggest moving to YAML optimized tasks AzureWebAppContainer(https://github.com/Microsoft/azure-pipelines-tasks/tree/master/Tasks/AzureWebAppContainerV1) and AzureFunctionAppContainer (https://github.com/Microsoft/azure-pipelines-tasks/tree/master/Tasks/AzureFunctionAppContainerV1). Though they are in preview, they use the same logic on the backend, with some cleanup. If the templates are in preview, it would be a safe and better choice to move to these new tasks.

File: arm-templates/acr.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (bansalaseem) [inline arm-templates/acr.json]: create a new folder under templates called resources and make sure that the file name is same as template name

File: arm-templates/webapp-on-containers.json [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+							"name": "DOCKER_REGISTRY_SERVER_USERNAME",
+							"value": "[listCredentials(concat('Microsoft.ContainerRegistry/registries/', parameters('registryName')), '2017-10-01').username]"
+						},
+						{
+							"name": "DOCKER_REGISTRY_SERVER_PASSWORD",
+							"value": "[listCredentials(concat('Microsoft.ContainerRegistry/registries/', parameters('registryName')), '2017-10-01').passwords[0].value]"
+						},
+						{
+							"name": "WEBSITES_ENABLE_APP_SERVICE_STORAGE",
+							"value": "false"
```
Comment (bansalaseem) [inline arm-templates/webapp-on-containers.json]: Remove this

File: arm-templates/webapp-on-containers.json [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+						}
+					],
+					"appCommandLine": "[parameters('startupCommand')]",
+					"linuxFxVersion": "[concat('DOCKER|', reference(concat('Microsoft.ContainerRegistry/registries/', parameters('registryName'))).loginServer, '/', parameters('imageName'))]"
+				},
+				"serverFarmId": "[concat('/subscriptions/', subscription().subscriptionId,'/resourcegroups/', resourceGroup().name, '/providers/Microsoft.Web/serverfarms/', parameters('hostingPlanName'))]",
+				"hostingEnvironment": ""
```
Comment (bansalaseem) [inline arm-templates/webapp-on-containers.json]: Remove this

File: arm-templates/webapp-on-containers.json [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+			"kind": "linux",
+			"name": "[parameters('hostingPlanName')]",
+			"apiVersion": "2016-09-01",
+			"location": "[resourceGroup().location]",
+			"properties": {
+				"name": "[parameters('hostingPlanName')]",
+				"workerSizeId": "0",
+				"reserved": true,
+				"numberOfWorkers": "1",
+				"hostingEnvironment": ""
```
Comment (bansalaseem) [inline arm-templates/webapp-on-containers.json]: Remove this

File: arm-templates/webapp-on-containers.json [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+				"Name": "[first(split(parameters('sku'), ' '))]"
+			},
+			"kind": "linux",
+			"name": "[parameters('hostingPlanName')]",
+			"apiVersion": "2016-09-01",
+			"location": "[resourceGroup().location]",
+			"properties": {
+				"name": "[parameters('hostingPlanName')]",
+				"workerSizeId": "0",
+				"reserved": true,
```
Comment (bansalaseem) [inline arm-templates/webapp-on-containers.json]: Remove this if its not required

File: arm-templates/functionapp.json [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                "networkAcls": {
+                    "bypass": "AzureServices",
+                    "virtualNetworkRules": [],
+                    "ipRules": [],
+                    "defaultAction": "Allow"
+                },
+                "supportsHttpsTrafficOnly": false,
+                "encryption": {
+                    "services": {
+                        "file": {
```
Comment (bansalaseem) [inline arm-templates/functionapp.json]: Remove all which are really not required.

File: arm-templates/functionapp.json [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+            }
+        },
+        {
+            "type": "Microsoft.Web/sites",
+            "name": "[parameters('functionAppName')]",
+            "kind": "functionapp",
+            "apiVersion": "2016-08-01",
+            "location": "[resourceGroup().location]",
+            "tags": {
+                "[concat('hidden-related:', '/subscriptions/', subscription().subscriptionId,'/resourcegroups/', resourceGroup().name, '/providers/Microsoft.Web/serverfarms/', parameters('hostingPlanName'))]": "empty"
```
Comment (bansalaseem) [inline arm-templates/functionapp.json]: Remove this

File: templates/resources/arm/aks.json [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+			"type": "string"
+		},
+		"servicePrincipalId": {
+			"type": "securestring"
+		},
+		"servicePrincipalKey": {
+			"type": "securestring"
+		},
+		"agentCount": {
+			"defaultValue": 3,
```
Comment (shashankbarsin) [inline templates/resources/arm/aks.json]: Should we set it to a lower count - 1 or 2?

File: templates/resources/arm/aks.json [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+			"type": "int"
+		},
+		"agentVMSize": {
+			"defaultValue": "Standard_D2_v2",
+			"type": "string"
+		},
+		"kubernetesVersion": {
+			"defaultValue": "1.11.5",
+			"type": "string"
```
Comment (bansalaseem) [inline templates/resources/arm/aks.json]: Shashank mentioned that we should try out by removing this and I like that idea.

File: templates/resources/arm/aks.json (hunk 8 above)
Comment (ds-ms) [reply templates/resources/arm/aks.json]: I've set it to 1 for now

File: templates/resources/arm/aks.json (hunk 9 above)
Comment (ds-ms) [reply templates/resources/arm/aks.json]: I removed it.

File: templates/docker-container-functionapp.yml [hunk 10]
```diff
@@ -59,45 +54,15 @@ jobs:
...
-          echo "Creating function app: $(FunctionApp.Name)"
-          az functionapp create -g $(Azure.ResourceGroupName) -n $(FunctionApp.Name) -s $(StorageAccount.Name) --plan $(ServicePlan.Name)
-          echo "Created function app: $(FunctionApp.Name)"
-        else
-          echo "Create Function App: false"
-        fi
+      azureSubscription: '$(Azure.ServiceConnectionId)'
+      resourceGroupName: '$(Azure.ResourceGroupName)'
+      location: '$(Azure.Location)'
+      templateLocation: 'URL of the file'
```
Comment (bansalaseem) [inline templates/docker-container-functionapp.yml]: create an issue on ARG task to not require "Url of the file" input

File: templates/docker-container-to-aks.yml [hunk 11]
```diff
@@ -54,48 +54,16 @@ jobs:
...
-            ACR_ID=$(az acr show --name $(ACR.Name) --resource-group $(Azure.ResourceGroupName) --query "id" --output tsv)
-            az role assignment create --assignee $CLIENT_ID --role Reader --scope $ACR_ID
-          fi
-        fi
-        AKS_ID=$(az aks show --resource-group $(Azure.ResourceGroupName) --name $(AKS.ClusterName) --query "id" -o tsv)
-        echo "AKS cluster: [$(AKS.ClusterName)](https://portal.azure.com/#resource${AKS_ID}/overview)" > $(System.DefaultWorkingDirectory)/azureCluster.md
-        echo "##vso[task.addattachment type=Distributedtask.Core.Summary;name=AzureResources;]$(System.DefaultWorkingDirectory)/azureCluster.md"
+      csmFileLink: 'https://raw.githubusercontent.com/Microsoft/azure-pipelines-yaml/master/templates/resources/arm/aks.json'
+      overrideParameters: '-registryName "$(ACR.Name)" -registryLocation "$(Azure.Location)" -servicePrincipalId $servicePrincipalId -servicePrincipalKey $servicePrincipalKey -clusterName "$(AKS.ClusterName)" -dnsPrefix "$(AKS.ClusterName)" -clusterLocation "$(Azure.Location)"'
```
Comment (bansalaseem) [inline templates/docker-container-to-aks.yml]: remove dnsPrefix

File: templates/resources/arm/aks.json [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+			"type": "securestring"
+		},
+		"agentCount": {
+			"defaultValue": 1,
+			"type": "int"
+		},
+		"agentVMSize": {
+			"defaultValue": "Standard_D2_v2",
+			"type": "string"
+		},
```
Comment (bansalaseem) [inline templates/resources/arm/aks.json]: Remove agentVMSize

File: templates/resources/arm/functionapp.json [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+            "kind": "Storage",
+            "name": "[parameters('storageAccountName')]",
+            "apiVersion": "2018-07-01",
+            "location": "[resourceGroup().location]",
+            "tags": {},
+            "scale": null,
+            "properties": {
+                "networkAcls": {
+                    "bypass": "AzureServices",
+                    "defaultAction": "Allow"
```
Comment (bansalaseem) [inline templates/resources/arm/functionapp.json]: Remove properties, scale, tags ...

File: templates/resources/arm/functionapp.json [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+                        {
+                            "name": "FUNCTIONS_WORKER_RUNTIME",
+                            "value": "[parameters('functionsRunTime')]"
+                        },
+                        {
+                            "name": "AzureWebJobsStorage",
+                            "value": "[concat('DefaultEndpointsProtocol=https;AccountName=',parameters('storageAccountName'),';AccountKey=',listKeys(resourceId('Microsoft.Storage/storageAccounts', parameters('storageAccountName')), '2015-05-01-preview').key1)]"
+                        },
+                        {
+                            "name": "FUNCTIONS_EXTENSION_VERSION",
```
Comment (bansalaseem) [inline templates/resources/arm/functionapp.json]: See whether we can remove it.

Comment (ds-ms) [reply templates/resources/arm/functionapp.json]: This is the Azure Function runtime version.

File: templates/resources/arm/aks.json (hunk 12 above)
Comment (ds-ms) [reply templates/resources/arm/aks.json]: I'll look into it.

**Pull request merged**